### ISSUE TITLE
Cannot restore volumes and stateful clusters with velero/restic

### ISSUE THREAD

**Issue by oldManLemon** [vmware-tanzu/velero#5006]
**What steps did you take and what happened:**
I ran `velero create backup my-database --include-namespaces database --default-volumes-to-restic` and got a completed message on the OG cluster. 
See logs 
```
Total items to be backed up:  20
Items backed up:              20

Resource List:
  apps/v1/ControllerRevision:
    - database/mariadb-primary-6c8864b4f5
  apps/v1/StatefulSet:
    - database/mariadb-primary
  discovery.k8s.io/v1beta1/EndpointSlice:
    - database/mariadb-primary-wlt7v
    - database/mariadb-secondary-pdlsn
  v1/ConfigMap:
    - database/kube-root-ca.crt
    - database/mariadb-primary
    - database/mariadb-secondary
  v1/Endpoints:
    - database/mariadb-primary
    - database/mariadb-secondary
  v1/Namespace:
    - database
  v1/PersistentVolume:
    - pvc-05ab194f-a3d8-4ee0-9199-b5f140fba4d1
    - pvc-0f6413f1-e8df-47f9-bca3-09bf10a639f8
  v1/PersistentVolumeClaim:
    - database/data-mariadb-primary-0
    - database/data-mariadb-secondary-0
  v1/Pod:
    - database/mariadb-primary-0
  v1/Secret:
    - database/default-token-5httx
    - database/mariadb
  v1/Service:
    - database/mariadb-primary
    - database/mariadb-secondary
  v1/ServiceAccount:
    - database/default

Velero-Native Snapshots: <none included>

Restic Backups:
  Completed:
    database/mariadb-primary-0: data
```

I then moved to my secondary cluster and attempted to restore the namespace. 

where at the end  of the restore decribe --details I get
 ```
Restore PVs:  auto

Restic Restores:
  Failed:
    database/mariadb-primary-0: data
```

Relevant info from the logs 
`level=error msg="unable to successfully complete restic restores of pod's volumes" error="pod volume restore failed: error restoring volume: error identifying path of volume: expected one matching path, got 0" logSource="pkg/restore/restore.go:1462" restore=velero/drew-database`

**What did you expect to happen:**

To get my volume restored as well, I can see it in the azure blob under restic. 

**The following information will help us better understand what's going on**:

_If you are using velero v1.7.0+:_  
Please use `velero debug  --backup <backupname> --restore <restorename>` to generate the support bundle, and attach to this issue, more options please refer to `velero debug --help` 

_If you are using earlier versions:_  
Please provide the output of the following commands (Pasting long output into a [GitHub gist](https://gist.github.com) or other pastebin is fine.)

https://pastebin.com/WCeuqCeS

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]
I am using a kind cluster to restore on. 

**Environment:**

- Velero version (use `velero version`): 

On restore machine
Client:
        Version: v1.5.1
        Git commit: 87d86a45a6ca66c6c942c7c7f08352e26809426c
Server:
        Version: v1.8.1

On backup machine

Client:
        Version: v1.5.3
        Git commit: 123109a3bcac11dbb6783d2758207bac0d0817cb
Server:
        Version: v1.5.3

- Velero features (use `velero client config get features`): 
features: <NOT SET>
- Kubernetes version (use `kubectl version`):

- Kubernetes installer & version:
- Cloud provider or hardware configuration:

vmware on backup

NAME="CentOS Stream"
VERSION="8"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="8"
PLATFORM_ID="platform:el8"
PRETTY_NAME="CentOS Stream 8"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:8"
HOME_URL="https://centos.org/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux 8"
REDHAT_SUPPORT_PRODUCT_VERSION="CentOS Stream"
restore on kind

- OS (e.g. from `/etc/os-release`):

centos

kind is debain

**Other**

I am fully willing to accept that have done this incorrectly but I have being trying everything for the past week to get even a single volume to restore using restic and have not succeeded. 

**Vote on this issue!**

This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "I would like to see this bug fixed as soon as possible"
- :-1: for "There are more important bugs to focus on right now"

**Comment by blackpiglet** [comment]
@oldManLemon 
I notice the versions of Veleros in backup and restore clusters are quite different.
Velero didn't test that case. Velero only makes sure minor one version difference working, e.g. 1.7~1.8
So could you make your two clusters' Velero version same, then take another try?

**Comment by oldManLemon** [comment]
Hi, I was actually hoping to use it as part of a migrating strategy, however it is just the restic backups that are not working. I will test with a matching cluster and post the results.

**Comment by oldManLemon** [comment]
OK I lowered it 1.20.7 where as the original is 1.20.6 as this what the kind repo had as stable image.

It is unfortunately still failing. Any insight would be appreciated 😊

Below is the output from decribe backup --details

https://pastebin.com/2SaPjqMc

And here is a log 

https://gist.github.com/oldManLemon/a67bed05fafb312f3b395fdd06ad5951

**Comment by oldManLemon** [comment]
Open

**Comment by blackpiglet** [comment]
@oldManLemon 
From the `velero restore describe` output, error informations are related to `pod volume restore failed: error restoring volume: error identifying path of volume: expected one matching path, got 0`

The error basically is caused by volumes in pod that are not mounted.  

This is an example Pod Spec yaml that will cause backup and restore fails due to the error.
We can see `volumes` has two items. `volumeMounts` has one item. The difference is a emptyDir type volume called `test`. Velero didn't check whether the volume is mounted or not during backup process, so it causes PartiallyFail. The is already fixed in Velero v1.9. 
You can either annotate your Pod to skip the non-mounted volumes, or use Velero v1.9, and please backup and restore again.

``` yaml
    spec:
      containers:
      - image: k8s.gcr.io/nginx-slim:0.8
        imagePullPolicy: IfNotPresent
        name: nginx
        ports:
        - containerPort: 80
          name: web
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /usr/share/nginx/html
          name: www
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
      volumes:
      - configMap:
          defaultMode: 420
          name: kube-root-ca.crt
        name: cm
      - emptyDir: {}
        name: test
```

**Comment by blackpiglet** [comment]
Closed for no response. Welcome to reopen, when your problem is still not resolved.
